File: custom-addons/morenitapp/models/libro.py
from odoo import models, fields, api

# Modelo Libro
# Gestiona libros/eventos/publicaciones

class Libro(models.Model):

    _name = 'morenitapp.libro'
    _description = 'Libros de la Hermandad'

    # Campo mostrado como nombre
    _rec_name = 'nombre'


    cod_libro = fields.Char(
        string="Código de Libro",
        required=True
    )

    nombre = fields.Char(
        string="Nombre del Libro",
        required=True
    )

    anio = fields.Integer(
        string="Año del Libro",
        required=True
    )

    descripcion = fields.Text(
        string="Descripción"
    )

    # DATOS ECONÓMICOS
    importe = fields.Float(
        string="Importe Total"
    )

    fecha_recibo = fields.Date(
        string="Fecha Recibo"
    )

    texto_recibo_evento = fields.Char(
        string="Texto del Recibo del Evento"
    )

    texto_anunciante = fields.Char(
        string="Texto del Anunciante del Libro"
    )

    # RELACIONES
    tipoevento_id = fields.Many2one(
        'morenitapp.tipoevento',
        string="Tipo de Evento"
    )

    # Archivos adjuntos
    adjuntos_ids = fields.Many2many(
        'ir.attachment',
        string="Documentos del Libro",
        help="Sube aquí todos los archivos que necesites"
    )

    # Lista de anunciantes
    anunciantes_ids = fields.One2many(
        'morenitapp.libroanunciante',
        'libro_id',
        string="Anunciantes del Libro"
    )

    # CAMPOS CALCULADOS
    total_anunciantes = fields.Float(
        string="Total Recaudado Anunciantes",
        compute="_compute_total_anunciantes",
        store=True
    )

    # SUMA TOTAL DE IMPORTES
    @api.depends('anunciantes_ids.importe')
    def _compute_total_anunciantes(self):

        for record in self:

            record.total_anunciantes = sum(
                record.anunciantes_ids.mapped('importe')
            )

File: mobile_flutter/morenitapp/lib/features/panel-gestion/libros/domain/entities/libro.dart
import 'libro_anunciante.dart';
import 'libro_adjunto.dart';


class Libro {
  final int id;
  final String codLibro;
  final String nombre;
  final int anio;
  final String descripcion;
  final double importe;
  final String? fechaRecibo;
  final int? tipoeventoId;
  final double totalAnunciantes;
  final List<LibroAnunciante> anunciantes;
  final List<LibroAdjunto> archivos;
  final String? textoReciboEvento;
  final String? textoAnunciante;

  Libro({
    required this.id,
    required this.codLibro,
    required this.nombre,
    required this.anio,
    required this.descripcion,
    required this.importe,
    this.fechaRecibo,
    this.tipoeventoId,
    this.totalAnunciantes = 0.0,
    this.anunciantes = const [],
    this.archivos = const [],
    this.textoAnunciante,
    this.textoReciboEvento,
  });

  factory Libro.fromJson(Map<String, dynamic> json) {
    return Libro(
      id: json['id'] is int ? json['id'] : int.parse(json['id'].toString()),
      codLibro: json['cod_libro'] ?? '',
      nombre: json['nombre'] ?? '',
      anio: json['anio'] ?? 0,
      descripcion: json['descripcion'] ?? '',
      importe: (json['importe'] ?? 0.0).toDouble(),
      fechaRecibo: json['fecha_recibo'],
      textoReciboEvento: json['texto_recibo_evento'],
      textoAnunciante: json['texto_anunciante'],
      tipoeventoId: json['tipoevento_id'],
      totalAnunciantes: (json['total_anunciantes'] ?? 0.0).toDouble(),
      anunciantes: (json['anunciantes'] as List? ?? [])
          .map((a) => LibroAnunciante.fromJson(a))
          .toList(),
      archivos: (json['archivos'] as List? ?? [])
          .map((f) => LibroAdjunto.fromJson(f))
          .toList(),
    );
  }
}
